syntax = "proto3";

enum cmsTagSignatureProto {
  CMSTagSignatureUnKnown                  = 0;           // The first enum value must be zero in proto3
  CMSSigAToB0Tag                          = 0x41324230;  // 'A2B0'
  CMSSigAToB1Tag                          = 0x41324231;  // 'A2B1'
  CMSSigAToB2Tag                          = 0x41324232;  // 'A2B2'
  CMSSigBlueColorantTag                   = 0x6258595A;  // 'bXYZ'
  CMSSigBlueMatrixColumnTag               = 0x6258595A;  // 'bXYZ'
  CMSSigBlueTRCTag                        = 0x62545243;  // 'bTRC'
  CMSSigBToA0Tag                          = 0x42324130;  // 'B2A0'
  CMSSigBToA1Tag                          = 0x42324131;  // 'B2A1'
  CMSSigBToA2Tag                          = 0x42324132;  // 'B2A2'
  CMSSigCalibrationDateTimeTag            = 0x63616C74;  // 'calt'
  CMSSigCharTargetTag                     = 0x74617267;  // 'targ'
  CMSSigChromaticAdaptationTag            = 0x63686164;  // 'chad'
  CMSSigChromaticityTag                   = 0x6368726D;  // 'chrm'
  CMSSigColorantOrderTag                  = 0x636C726F;  // 'clro'
  CMSSigColorantTableTag                  = 0x636C7274;  // 'clrt'
  CMSSigColorantTableOutTag               = 0x636C6F74;  // 'clot'
  CMSSigColorimetricIntentImageStateTag   = 0x63696973;  // 'ciis'
  CMSSigCopyrightTag                      = 0x63707274;  // 'cprt'
  CMSSigCrdInfoTag                        = 0x63726469;  // 'crdi'
  CMSSigDataTag                           = 0x64617461;  // 'data'
  CMSSigDateTimeTag                       = 0x6474696D;  // 'dtim'
  CMSSigDeviceMfgDescTag                  = 0x646D6E64;  // 'dmnd'
  CMSSigDeviceModelDescTag                = 0x646D6464;  // 'dmdd'
  CMSSigDeviceSettingsTag                 = 0x64657673;  // 'devs'
  CMSSigDToB0Tag                          = 0x44324230;  // 'D2B0'
  CMSSigDToB1Tag                          = 0x44324231;  // 'D2B1'
  CMSSigDToB2Tag                          = 0x44324232;  // 'D2B2'
  CMSSigDToB3Tag                          = 0x44324233;  // 'D2B3'
  CMSSigBToD0Tag                          = 0x42324430;  // 'B2D0'
  CMSSigBToD1Tag                          = 0x42324431;  // 'B2D1'
  CMSSigBToD2Tag                          = 0x42324432;  // 'B2D2'
  CMSSigBToD3Tag                          = 0x42324433;  // 'B2D3'
  CMSSigGamutTag                          = 0x67616D74;  // 'gamt'
  CMSSigGrayTRCTag                        = 0x6b545243;  // 'kTRC'
  CMSSigGreenColorantTag                  = 0x6758595A;  // 'gXYZ'
  CMSSigGreenMatrixColumnTag              = 0x6758595A;  // 'gXYZ'
  CMSSigGreenTRCTag                       = 0x67545243;  // 'gTRC'
  CMSSigLuminanceTag                      = 0x6C756d69;  // 'lumi'
  CMSSigMeasurementTag                    = 0x6D656173;  // 'meas'
  CMSSigMediaBlackPointTag                = 0x626B7074;  // 'bkpt'
  CMSSigMediaWhitePointTag                = 0x77747074;  // 'wtpt'
  CMSSigNamedColorTag                     = 0x6E636f6C;  // 'ncol' // Deprecated by the ICC
  CMSSigNamedColor2Tag                    = 0x6E636C32;  // 'ncl2'
  CMSSigOutputResponseTag                 = 0x72657370;  // 'resp'
  CMSSigPerceptualRenderingIntentGamutTag = 0x72696730;  // 'rig0'
  CMSSigPreview0Tag                       = 0x70726530;  // 'pre0'
  CMSSigPreview1Tag                       = 0x70726531;  // 'pre1'
  CMSSigPreview2Tag                       = 0x70726532;  // 'pre2'
  CMSSigProfileDescriptionTag             = 0x64657363;  // 'desc'
  CMSSigProfileDescriptionMLTag           = 0x6473636d;  // 'dscm'
  CMSSigProfileSequenceDescTag            = 0x70736571;  // 'pseq'
  CMSSigProfileSequenceIdTag              = 0x70736964;  // 'psid'
  CMSSigPs2CRD0Tag                        = 0x70736430;  // 'psd0'
  CMSSigPs2CRD1Tag                        = 0x70736431;  // 'psd1'
  CMSSigPs2CRD2Tag                        = 0x70736432;  // 'psd2'
  CMSSigPs2CRD3Tag                        = 0x70736433;  // 'psd3'
  CMSSigPs2CSATag                         = 0x70733273;  // 'ps2s'
  CMSSigPs2RenderingIntentTag             = 0x70733269;  // 'ps2i'
  CMSSigRedColorantTag                    = 0x7258595A;  // 'rXYZ'
  CMSSigRedMatrixColumnTag                = 0x7258595A;  // 'rXYZ'
  CMSSigRedTRCTag                         = 0x72545243;  // 'rTRC'
  CMSSigSaturationRenderingIntentGamutTag = 0x72696732;  // 'rig2'
  CMSSigScreeningDescTag                  = 0x73637264;  // 'scrd'
  CMSSigScreeningTag                      = 0x7363726E;  // 'scrn'
  CMSSigTechnologyTag                     = 0x74656368;  // 'tech'
  CMSSigUcrBgTag                          = 0x62666420;  // 'bfd '
  CMSSigViewingCondDescTag                = 0x76756564;  // 'vued'
  CMSSigViewingConditionsTag              = 0x76696577;  // 'view'
  CMSSigVcgtTag                           = 0x76636774;  // 'vcgt'
  CMSSigMetaTag                           = 0x6D657461;  // 'meta'
  CMSSigcicpTag                           = 0x63696370;  // 'cicp'
  CMSSigArgyllArtsTag                     = 0x61727473;  // 'arts'
  CMSSigMHC2Tag                           = 0x4D484332;   // 'MHC2'
  option allow_alias = true;
}

message tmStructProto {
  int32 tm_sec = 1;   // Seconds. [0-60] (1 leap second)
  int32 tm_min = 2;   // Minutes. [0-59]
  int32 tm_hour = 3;  // Hours. [0-23]
  int32 tm_mday = 4;  // Day. [1-31]
  int32 tm_mon = 5;   // Month. [0-11]
  int32 tm_year = 6;  // Year - 1900.
  int32 tm_wday = 7;  // Day of week. [0-6]
  int32 tm_yday = 8;  // Days in year. [0-365]
  int32 tm_isdst = 9; // DST. [-1/0/1]
  // int64 tm_gmtoff = 10;   // Seconds east of UTC.
  // string tm_zone = 11;    // Timezone abbreviation.
}

enum cmsProfileClassSignatureProto{
  CMSProfileClassSignatureUnKnown         = 0;           // The first enum value must be zero in proto3
  CMSSigInputClass                        = 0x73636E72;  // 'scnr'
  CMSSigDisplayClass                      = 0x6D6E7472;  // 'mntr'
  CMSSigOutputClass                       = 0x70727472;  // 'prtr'
  CMSSigLinkClass                         = 0x6C696E6B;  // 'link'
  CMSSigAbstractClass                     = 0x61627374;  // 'abst'
  CMSSigColorSpaceClass                   = 0x73706163;  // 'spac'
  CMSSigNamedColorClass                   = 0x6e6d636c;  // 'nmcl'

}

enum cmsColorSpaceSignatureProto{
  CMSColorSpaceSignatureUnKnown           = 0;           // The first enum value must be zero in proto3
  CMSSigXYZData                           = 0x58595A20;  // 'XYZ '
  CMSSigLabData                           = 0x4C616220;  // 'Lab '
  CMSSigLuvData                           = 0x4C757620;  // 'Luv '
  CMSSigYCbCrData                         = 0x59436272;  // 'YCbr'
  CMSSigYxyData                           = 0x59787920;  // 'Yxy '
  CMSSigRgbData                           = 0x52474220;  // 'RGB '
  CMSSigGrayData                          = 0x47524159;  // 'GRAY'
  CMSSigHsvData                           = 0x48535620;  // 'HSV '
  CMSSigHlsData                           = 0x484C5320;  // 'HLS '
  CMSSigCmykData                          = 0x434D594B;  // 'CMYK'
  CMSSigCmyData                           = 0x434D5920;  // 'CMY '
  CMSSigMCH1Data                          = 0x4D434831;  // 'MCH1'
  CMSSigMCH2Data                          = 0x4D434832;  // 'MCH2'
  CMSSigMCH3Data                          = 0x4D434833;  // 'MCH3'
  CMSSigMCH4Data                          = 0x4D434834;  // 'MCH4'
  CMSSigMCH5Data                          = 0x4D434835;  // 'MCH5'
  CMSSigMCH6Data                          = 0x4D434836;  // 'MCH6'
  CMSSigMCH7Data                          = 0x4D434837;  // 'MCH7'
  CMSSigMCH8Data                          = 0x4D434838;  // 'MCH8'
  CMSSigMCH9Data                          = 0x4D434839;  // 'MCH9'
  CMSSigMCHAData                          = 0x4D434841;  // 'MCHA'
  CMSSigMCHBData                          = 0x4D434842;  // 'MCHB'
  CMSSigMCHCData                          = 0x4D434843;  // 'MCHC'
  CMSSigMCHDData                          = 0x4D434844;  // 'MCHD'
  CMSSigMCHEData                          = 0x4D434845;  // 'MCHE'
  CMSSigMCHFData                          = 0x4D434846;  // 'MCHF'
  CMSSigNamedData                         = 0x6e6d636c;  // 'nmcl'
  CMSSig1colorData                        = 0x31434C52;  // '1CLR'
  CMSSig2colorData                        = 0x32434C52;  // '2CLR'
  CMSSig3colorData                        = 0x33434C52;  // '3CLR'
  CMSSig4colorData                        = 0x34434C52;  // '4CLR'
  CMSSig5colorData                        = 0x35434C52;  // '5CLR'
  CMSSig6colorData                        = 0x36434C52;  // '6CLR'
  CMSSig7colorData                        = 0x37434C52;  // '7CLR'
  CMSSig8colorData                        = 0x38434C52;  // '8CLR'
  CMSSig9colorData                        = 0x39434C52;  // '9CLR'
  CMSSig10colorData                       = 0x41434C52;  // 'ACLR'
  CMSSig11colorData                       = 0x42434C52;  // 'BCLR'
  CMSSig12colorData                       = 0x43434C52;  // 'CCLR'
  CMSSig13colorData                       = 0x44434C52;  // 'DCLR'
  CMSSig14colorData                       = 0x45434C52;  // 'ECLR'
  CMSSig15colorData                       = 0x46434C52;  // 'FCLR'
  CMSSigLuvKData                          = 0x4C75764B;   // 'LuvK'
}

enum cmsPlatformSignatureProto{
  CMSPlatformSignatureUnKnown           = 0;           // The first enum value must be zero in proto3
  CMSSigMacintosh                         = 0x4150504C;  // 'APPL'
  CMSSigMicrosoft                         = 0x4D534654;  // 'MSFT'
  CMSSigSolaris                           = 0x53554E57;  // 'SUNW'
  CMSSigSGI                               = 0x53474920;  // 'SGI '
  CMSSigTaligent                          = 0x54474E54;  // 'TGNT'
  CMSSigUnices                            = 0x2A6E6978;   // '*nix'   // From argyll -- Not official
} 

message cmsProfileIDProto {
  repeated uint32 id32 = 1;
  // repeated uint32 id16 = 2;
  // repeated uint32 id8 = 3;
}

message cmsICCPROFILEProto{
  // cmsIOHANDLER*           IOhandler;
  // cmsContext              ContextID;
  tmStructProto                           Created = 3;
  uint32                                  CMM = 4;
  uint32                                  Version = 5;
  cmsProfileClassSignatureProto           DeviceClass = 6;
  cmsColorSpaceSignatureProto             ColorSpace = 7;
  cmsColorSpaceSignatureProto             PCS = 8;
  uint32                                  RenderingIntent = 9;
  cmsPlatformSignatureProto               platform  = 10;
  uint32                                  flags = 11;
  uint32                                  manufacturer = 12;
  uint32                                  model = 13;
  uint64                                  attributes = 14;
  uint32                                  creator = 15;
  cmsProfileIDProto                       ProfileID = 16; 
  uint32                                  TagCount = 17;
  repeated cmsTagSignatureProto           TagNames = 18;
  repeated cmsTagSignatureProto           TagLinked = 19;           // The tag to which is linked (0=none)
  repeated uint32                         TagSizes = 20;            // Size on disk
  repeated uint32                         TagOffsets = 21;
  repeated int32                          TagSaveAsRaw = 22;        // True to write uncooked
  repeated bytes                          tag_ptrs = 23; 
  // cmsTagTypeHandler*                      TagTypeHandlers[MAX_TABLE_TAG];     // Same structure may be serialized on different types
  //                                                                             // depending on profile version, so we keep track of the
  //                                                                             // type handler for each tag in the list.
  // Special
  // int32                                     IsWrite = 25;
  // Keep a mutex for cmsReadTag -- Note that this only works if the user includes a mutex plugin
  // void *                   UsrMutex;
}
